Change 'actor class' to 'actor'

etcwilde opened the pull request:
We've decided to go with calling actor classes `actor` instead of `actor class`.
We still want to be able to parse in `actor class` and recommend the correct fix-its to make it just `actor`. This is a little challenging since we parse other modifiers in any order, so `actor public class Foo` and `public actor class Foo` were both totally parsable. We still need to be able to do that, which makes this change less pleasant.

Furthermore, we want to make it a contextual keyword instead of an actual keyword so that we avoid breaking as much source. This also adds to the complexity.

I will update the diagnostics to refer to `actor` later.

Resolves: rdar://73728503

ktoso commented on test/decl/class/actor/basic.swift:
```diff
@@ -2,12 +2,20 @@
 class NonActor { }
 
-actor class NonActorSubclass : NonActor { } // expected-error{{actor class cannot inherit from non-actor class 'NonActor'}}
+actor NonActorSubclass : NonActor { } // expected-error{{actor class cannot inherit from non-actor class 'NonActor'}}
```
should we instead be calling them "actor" rather than actor class in all diagnostics from here on?

etcwilde replied:
Yeah, that seems like a good idea.

etcwilde committed:
Rename 'actor class' -> 'actor'

This patch softly updates the spelling of actors from `actor class` to
`actor`. We still accept using `actor` as a modifying attribute of
class, but emit a warning and fix-it to make the change.

One of the challenges that makes this messier is that the modifier list
can be in any order. e.g, `public actor class Foo {}` is the same as
`actor public class Foo {}`.

Classes have been updated to include whether they were explicitly
declared as an actor. This change updates the swiftmodule serialization
version number to 0.591. The additional bit only gets set of the class
declaration was declared as an actor, not if the actor was applied as an
attribute. This allows us to correctly emit `actor class` vs `actor`
emitting the code back out.

etcwilde committed:
Update standard library actor class

This patch updates the once instance of actor class in the concurrency
library to actor. This is the actor for the MainActor global actor.

etcwilde committed:
Updating tests to use actor

This patch updates the `actor class` spelling to `actor` in almost all
of the tests. There are places where I verify that we sanely handle
`actor` as an attribute though. These include:

 - test/decl/class/actor/basic.swift
 - test/decl/protocol/special/Actor.swift
 - test/SourceKit/CursorInfo/cursor_info_concurrency.swift
 - test/attr/attr_objc_async.swift
 - test/ModuleInterface/actor_protocol.swift

etcwilde commented:
@swift-ci please test

DougGregor commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -4148,6 +4206,13 @@ Parser::parseDecl(ParseDeclOptions Flags,
   // Obvious nonsense.
   default:
 
+    if (shouldParseExperimentalConcurrency() && Tok.is(tok::identifier) &&
```
You don't need the `Tok.is(too::identifier)` here, because it's part of `Tok.isContextualKeyword`.

DougGregor commented on include/swift/AST/Decl.h:
```diff
@@ -3587,7 +3590,8 @@ class ClassDecl final : public NominalTypeDecl {
             ArrayRef<TypeLoc> Inherited,
-            GenericParamList *GenericParams, DeclContext *DC);
+            GenericParamList *GenericParams, DeclContext *DC,
+            bool isActor=false);
```
It might be worth dropping the default argument so we don't forget to pass it somewhere in the future.

etcwilde replied:
Dropping it got a little nastier than intended. The unit tests have a `makeNominal` that will construct something based on a template argument. That thing is either a `ClassDecl` or a `StructDecl`.

Ended up with this little specialization of `makeNominal`. Not great, but not terrible.
```c++
  template <typename Nominal>
  Nominal *makeNominal(StringRef name,
                       GenericParamList *genericParams = nullptr) {

    auto result = new (Ctx) Nominal(SourceLoc(), Ctx.getIdentifier(name),
                                    SourceLoc(), /*inherited*/{},
                                    genericParams, FileForLookups);
    result->setAccess(AccessLevel::Internal);
    return result;
  }

  template <>
  swift::ClassDecl *makeNominal(StringRef name,
                                GenericParamList *genericParams) {
    auto result = new (Ctx) ClassDecl(SourceLoc(), Ctx.getIdentifier(name),
                                      SourceLoc(), /*inherited*/{},
                                      genericParams, FileForLookups,
                                      /*isActor*/false);
    result->setAccess(AccessLevel::Internal);
    return result;
  }
```

DougGregor reviewed (approved):
This looks great; just two trivial comments that shouldn't block you from merging.

etcwilde commented:
@swift-ci please test

etcwilde commented:
Heh, there were more places where constructed ClassDecls. :) I think I have all of them fixed, just testing locally.

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test
